pkg/backplane/ca/ca.go

```diff
@@ -0,0 +1,176 @@
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"math/big"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	rsaBits  = 2048
+	validFor = 48 * time.Hour
+	caName   = "sonobuoy-ca"
+)
+
+var (
+	pkixName = pkix.Name{
+		Organization:       []string{"Heptio"},
+		OrganizationalUnit: []string{"sonobuoy"},
+		Country:            []string{"USA"},
+		Locality:           []string{"Seattle"},
+	}
+
+	randReader = rand.Reader
+)
+
+// Authority represents a root certificate authority that can issues
+// certificates to be used for Client certs.
+// Sonobuoy issues every worker a client certificate
+type Authority struct {
+	privKey    *rsa.PrivateKey
+	cert       *x509.Certificate
+	lastSerial *big.Int
+}
+
+// NewAuthority creates a new certificate authority. A new private key and root certificate will
+// be generated but not returned.
+func NewAuthority() (*Authority, error) {
+
+	privKey, err := rsa.GenerateKey(randReader, rsaBits)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't generate private key")
+	}
+	auth := &Authority{
+		privKey: privKey,
+	}
+	cert, err := auth.makeCert(privKey.Public(), func(cert *x509.Certificate) {
+		cert.IsCA = true
+		cert.KeyUsage = x509.KeyUsageCertSign
+		cert.Subject.CommonName = caName
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't create certificate authority root certificate")
+	}
+	auth.cert = cert
+	return auth, nil
+}
+
+// makeCert takes a public key and a function to mutate the certificate template with updated parameters
+func (a *Authority) makeCert(pub crypto.PublicKey, mut func(*x509.Certificate)) (*x509.Certificate, error) {
+
+	serialNumber := a.nextSerial()
+	validFrom := time.Now()
+	tmpl := x509.Certificate{
+		SerialNumber:          serialNumber,
+		Subject:               pkixName,
+		NotBefore:             validFrom,
+		NotAfter:              validFrom.Add(validFor),
+		KeyUsage:              0,
+		ExtKeyUsage:           []x509.ExtKeyUsage{},
+		BasicConstraintsValid: true,
+	}
+	mut(&tmpl)
+	parent := a.cert
+	// NewAuthority case
+	if a.cert == nil {
+		parent = &tmpl
+	}
+
+	newDERCert, err := x509.CreateCertificate(randReader, &tmpl, parent, pub, a.privKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't make new certificate")
+	}
+
+	cert, err := x509.ParseCertificate(newDERCert)
+	return cert, errors.Wrap(err, "couldn't re-parse created certificate")
+}
+
+func (a *Authority) makeLeafCert(mut func(*x509.Certificate)) (*tls.Certificate, error) {
+	privKey, err := rsa.GenerateKey(randReader, rsaBits)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't generate private key")
+	}
+
+	cert, err := a.makeCert(privKey.Public(), mut)
+
+	return &tls.Certificate{
+		Certificate: [][]byte{cert.Raw, a.cert.Raw},
+		PrivateKey:  privKey,
+		Leaf:        cert,
+	}, errors.Wrap(err, "couldn't make leaf cert")
+
+}
+
+func (a *Authority) nextSerial() *big.Int {
+	if a.lastSerial == nil {
```

Comment:
Not sure if it matters for the intended use case, but this isn't thread safe. It might be nice to add a `sync.Mutex` or add an explicit warning that `ServerKeyPair`/`MakeServerConfig ` cannot be called on the same `Authority` from multiple goroutines.